### PR TITLE
Code improvements

### PR DESCRIPTION
**Pull request by pawelkg**
Partly fix issue thephpleague/uri-src#29

### CODE REVIEW

File: src/Http.php [hunk 1]
```diff
@@ -21,6 +21,11 @@
 use function method_exists;
 use function sprintf;
 
+/**
+ * Class Http.
```
Comment (nyamsprod) [inline src/Http.php:25]: this dock block is not really needed. I used to add it but I removed it long ago as it provides not added value at all to the library or its usage. To get more info about one class its better to refer to the documentation or the source code directly

File: src/Http.php [hunk 2]
```diff
@@ -45,11 +50,13 @@ private function __construct(UriInterface $uri)
     private function validate(UriInterface $uri): void
     {
         $scheme = $uri->getScheme();
+
```
Comment (nyamsprod) [inline src/Http.php:53]: Please remove all the extra empty line added. Put them in a separate PR since it only convey coding style change. And FWIW I'm not in favor of this change. You can change and or update your IDE settings to get more or less spaces between lines and get the same result in the end.

File: src/UriInfo.php [hunk 3]
```diff
@@ -22,8 +22,16 @@
...
 
+/**
+ * Class UriInfo.
+ *
+ * @package League\Uri
+ */
 final class UriInfo
 {
+    /**
+     * Regular expression pattern to find encoded chars.
```
Comment (nyamsprod) [inline src/UriInfo.php:33]: I have mix feeling about this comment/docblock . The constant name is explicit enough to not have to add this information. I'm only adding comment where necessary and I don't think it is necessary in this case.

File: src/UriResolver.php [hunk 4]
```diff
@@ -29,10 +29,15 @@
...
+/**
+ * Class UriResolver.
+ *
+ * @package League\Uri
+ */
 final class UriResolver
 {
     /**
-     * @var array
+     * Array of segments with dot(s).
```
Comment (nyamsprod) [inline src/UriResolver.php:40]: Comment not needed

File: src/UriString.php [hunk 5]
```diff
@@ -60,15 +41,21 @@
...
+        'host' => null,
+        'port' => null,
+        'path' => '',
+        'query' => null,
+        'fragment' => null,
     ];
 
     /**
-     * simple URI which do not need any parsing.
+     * sSimple URI which do not need any parsing.
```
Comment (nyamsprod) [inline src/UriString.php:58]: `sSimple` typo.

File: src/Uri.php [hunk 6]
```diff
@@ -254,16 +219,8 @@ final class Uri implements UriInterface
...
-        ?string $scheme,
-        ?string $user,
-        ?string $pass,
-        ?string $host,
-        ?int $port,
-        string $path,
-        ?string $query,
-        ?string $fragment
-    ) {
+    private function __construct(?string $scheme, ?string $user, ?string $pass, ?string $host, ?int $port, string $path, ?string $query, ?string $fragment)
```
Comment (nyamsprod) [inline src/Uri.php:222]: Please remove coding style. Here you are even violating PSR-2 coding style. When there are too many parameters its better for readability to align them on multiple lines.

File: src/Uri.php [hunk 7]
```diff
@@ -536,8 +407,7 @@ public static function __set_state(array $components): self
      *
      * The returned URI must be absolute.
      *
-     * @param mixed $uri      the input URI to create
```
Comment (nyamsprod) [inline src/Uri.php:539]: Why remove the docblock this time ? The variable are commented because they are mixed.

File: src/Uri.php [hunk 8]
```diff
@@ -623,20 +493,22 @@ public static function createFromComponents(array $components = []): self
     /**
      * Create a new instance from a data file path.
      *
-     * @param resource|null $context
```
Comment (nyamsprod) [inline src/Uri.php:626]: `resource` is a pseudo type in PHP no need to change this docblock

File: src/Common.php [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+
+        if ($uft8) {
+            $host = strtolower($host);
+            $convert = idn_to_utf8(
+                $host,
+                IDNA_CHECK_BIDI | IDNA_CHECK_CONTEXTJ | IDNA_NONTRANSITIONAL_TO_UNICODE,
+                INTL_IDNA_VARIANT_UTS46,
+                $idna_info
+            );
+        } else {
```
Comment (nyamsprod) [inline src/Common.php:167]: I'm against the use of `else` I prefer returning early returning often. It's even easier then to test and to maintain

File: src/Common.php [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+        return $format ? $idn : $host;
+    }
+
+    /**
+     * Filter IDN domain to UTF8 or ASCII.
+     *
+     * @throws SyntaxError
+     * @throws IdnSupportMissing
+     */
+    private static function checkIDN(string $host, bool $uft8 = true): string
```
Comment (nyamsprod) [inline src/Common.php:155]: I'm against using `boolean` parameters whatever the case. This method should be changed so that no boolean is needed.

File: src/Common.php [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace League\Uri;
+
+use League\Uri\Exceptions\IdnSupportMissing;
+use League\Uri\Exceptions\SyntaxError;
+
+final class Common
```
Comment (nyamsprod) [inline src/Common.php:19]: This class exposes too much implementation details to the public API which restricts changing that implementation during a minor release if necessary (This case has happen in v5x series and I want to avoid this burden on the new version). Turning the class into an abstract one won't change that fact. Sometimes duplication code is better than tightling coupling different part of the same library.

File: src/UriString.php [hunk 12]
```diff
@@ -259,12 +204,38 @@ public static function parse($uri): array
...
+            ]
+        );
+    }
+
+    /**
+     * Parse URI when it can be simply parsed.
+     *
+     * @throws SyntaxError
+     */
+    private static function parseSimply(string $uri): ?array
```
Comment (nyamsprod) [inline src/UriString.php:232]: What's the added value of declaring this method ? `parseSimply`

File: src/UriString.php [hunk 13]
```diff
@@ -291,6 +262,16 @@ public static function parse($uri): array
...
 
+        return null;
+    }
+
+    /**
+     * Check scheme and path parts of uri.
+     *
+     * @throws SyntaxError
+     */
+    private static function checkSchemePath(string $uri): array
```
Comment (nyamsprod) [inline src/UriString.php:273]: What's the added value of adding `checkSchemePath` ?

File: src/Uri.php [hunk 14]
```diff
@@ -52,100 +48,69 @@
...
-
+    /**
+     * Regular expression pattern to for file URI.
+     *
+     * @var string
+     */
     private const REGEXP_FILE_PATH = ',^(?<delim>/)?(?<root>[a-zA-Z][:|\|])(?<rest>.*)?,';
 
+    /**
+     * Mimetype retular expression pattern.
```
Comment (nyamsprod) [inline src/Uri.php:94]: typo `retular` to `regular`

**Review by nyamsprod — COMMENTED** [review]
This is a preliminary comment on the PR. Maybe to better approach it you should remove any coding style change and only focus on the changes you would like to submit on the code flow.

FWIW I'm not really in favor of the `Common` class addition as it exposes most of its API through static public methods.

These methods represent implementation details and therefore should not be expose in any shape of form to the outside world to improve code fixing and code modification in the future.

Also keep in mind that the code is a trade-off between readability, maintenance and performance (speed).

**Comment by pawelkg** [comment]
Thanks for feedback. I will close this PR and prepare separate ones. If idea of Common class not find Yours approvement, please tell me if changing it into trait will be more acceptable?

File: src/UriString.php (hunk 12 above)
Comment (pawelkg) [reply src/UriString.php:232]: Same as checkSchemePath I intended to make code more readable.

File: src/Common.php (hunk 10 above)
Comment (pawelkg) [reply src/Common.php:155]: OK. I will think about other implementation.

**Comment by pawelkg** [comment]
New PR is thephpleague/uri#133
